src/middlewared/middlewared/alert/source/quota.py

```diff
@@ -0,0 +1,147 @@
+
+logger = logging.getLogger(__name__)
+
+
+class QuotaAlertSource(ThreadedAlertSource):
+    level = AlertLevel.WARNING
+    title = "Quota exceed on dataset"
+
+    schedule = IntervalSchedule(timedelta(minutes=5))
+
+    def check_sync(self):
+        alerts = []
+
+        quotas = {}
+        for dataset in self.middleware.call_sync("datastore.query", "storage.dataset"):
+            quotas[dataset["name"]] = dataset
+        quotas[""] = self.middleware.call_sync("datastore.query", "storage.rootdataset", [], {"get": True})
+
+        with libzfs.ZFS() as zfs:
+            datasets = [
+                {
+                    k: v.__getstate__()
+                    for k, v in i.properties.items()
+                    if k in ["name", "quota", "used", "refquota", "usedbydataset", "mounted", "mountpoint"]
+                }
+                for i in zfs.datasets
+            ]
+
+        for d in datasets:
+            d["name"] = d["name"]["rawvalue"]
+
+        datasets = sorted(datasets, key=lambda ds: ds["name"])
+
+        for dataset in datasets:
+            for quota_property, used_property in [
+                ("quota", "used"),
+                ("refquota", "usedbydataset"),
+            ]:
+                try:
+                    quota_value = int(dataset[quota_property]["rawvalue"])
+                except (AttributeError, KeyError, ValueError):
+                    continue
+
+                if quota_value == 0:
+                    continue
+
+                used = int(dataset[used_property]["rawvalue"])
+                try:
+                    used_fraction = used / quota_value
+                except ZeroDivisionError:
+                    used_fraction = 1.0
+
+                critical_threshold = self._get_quota(quotas, dataset["name"], f"{quota_property}_critical")
+                warning_threshold = self._get_quota(quotas, dataset["name"], f"{quota_property}_warning")
+                if critical_threshold != 0 and used_fraction >= critical_threshold:
+                    level = AlertLevel.CRITICAL
+                elif warning_threshold != 0 and used_fraction >= warning_threshold:
+                    level = AlertLevel.WARNING
+                else:
+                    continue
+
+                quota_name = quota_property[0].upper() + quota_property[1:]
+
+                hostname = socket.gethostname()
+
+                title = ("%(name)s exceed on dataset %(dataset)s. "
+                         "Used %(percent_used).2f%% (%(used)s of %(quota_value)s)")
+                args = {
+                    "name": quota_name,
+                    "dataset": dataset["name"],
+                    "percent_used": used_fraction * 100,
+                    "used": humanfriendly.format_size(used),
+                    "quota_value": humanfriendly.format_size(quota_value),
+                }
+
+                mail = None
+                owner = self._get_owner(dataset)
+                if owner != 0:
```

Comment:
So we dont send email if the owner is root? Why is that?

Reply:
root is supposed to receive an e-mail with all alerts update, this is extra e-mail for specific user.